### PR TITLE
add compilation stage to dotnet pipelines

### LINKED ISSUE #13: Add Compilation stage to dotnet pipelines
# Description

The PR pipeline for dotnet should build the C# and fail if there are build errors

# Acceptance Criteria
- PR Workflow modified to include compilation step

### PR DESCRIPTION
**Pull request by polatengin**
# Description

Adding the following steps to pr_dotnet pipeline;

- setup environment for dotnet 6
- restore dependencies from nuget
- build dotnet projects against release configuration

_Note: MegaLinter step fails other than Linux agents, so, pr_dotnet.yml#L24 has only `ubuntu-latest` agent in the matrix. If we can run MegaLinter step on Windows and MacOS too, we can add those OSs to the matrix list_

Fixes #13 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] GitHub Action executed, logs can be found https://github.com/benchpress-team/benchpress/actions/runs/3316754200

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by ThePedestrian — APPROVED** [review]

File: BenchPress.sln
Comment (DilmurodMak) [inline BenchPress.sln:16]: Why the BenchPress.Module folder is deleted from project?

Comment (polatengin) [reply BenchPress.sln:16]: because that project does not exist 🙂

**Comment by DilmurodMak** [comment]
I thought I add it on dotnet folder structure.  I will make sure to added it

.
├── engine
│   ├── BenchPress.TestEngine
│   └── BenchPress.TestEngine.Tests
└── framework
│    └── dotnet
│        ├── BenchPress.TestFramework
│        ├── BenchPress.TestFramework.Tests
│        └── BenchPress.Module
└── samples
     └── dotnet
         └── BenchPress.Sample

**Comment by polatengin** [comment]
> I thought I add it on dotnet folder structure. I will make sure to added it
> 
> . ├── engine │ ├── BenchPress.TestEngine │ └── BenchPress.TestEngine.Tests └── framework │ └── dotnet │ ├── BenchPress.TestFramework │ ├── BenchPress.TestFramework.Tests │ └── BenchPress.Module └── samples └── dotnet └── BenchPress.Sample

you can add it to this branch, or, branch off-of this branch and create a PR, or, wait for this PR to be merged, then add it.

what you think @DilmurodMak?

**Comment by polatengin** [comment]
@Rhernandez513 , it seems github action fails because the content of CONTRIBUTE.md file in the root (https://github.com/benchpress-team/benchpress/actions/runs/3316951186/jobs/5479305221#step:4:254)

can you double check if this is the expected content for this file, please?

**Review by DilmurodMak — APPROVED** [review]

**Comment by Rhernandez513** [comment]
There is a warning that could be a potential cybersecurity concern and I would block the PR

[lint-ubuntu-latest](https://github.com/benchpress-team/benchpress/actions/runs/3316754200/jobs/5478906413)
Node.js 12 actions are deprecated. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/. Please update the following actions to use Node.js 16: actions/setup-dotnet

It looks like our actions are using Node 12, we should point it to Node 16 LTS `v16.18.0` is the current LTS at the time of writing

**Comment by Rhernandez513** [comment]
There is another Warning  at https://github.com/benchpress-team/benchpress/actions/runs/3316951186/jobs/5479305221#step:4:254

Warning: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

This one is non-blocking, but should be remedied as soon as convenient.

**Comment by polatengin** [comment]
> There is a warning that could be a potential cybersecurity concern and I would block the PR
> 
> [lint-ubuntu-latest](https://github.com/benchpress-team/benchpress/actions/runs/3316754200/jobs/5478906413) Node.js 12 actions are deprecated. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/. Please update the following actions to use Node.js 16: actions/setup-dotnet
> 
> It looks like our actions are using Node 12, we should point it to Node 16 LTS `v16.18.0` is the current LTS at the time of writing

we don't explicitly install or use Node in this project. I think GitHub hosted pipeline agents (we're using `ubuntu-latest`) has few of Node versions installed to be backward compatible, including Node v12.

Since we're not using it, and it's coming with the GitHub Actions platform, I don't think we should do anything about it.

Possible resolution may be creating our own GitHub Action agent, installing and maintaining all the sdks, libraries, tools, vs. on it. Which I'm skeptical about.

what team think? @DilmurodMak @Rhernandez513 @ThePedestrian @omusavi @jessica-ern

**Comment by Rhernandez513** [comment]
@polatengin 

try this config to fix the spell linting error:

```json 
{
    "ignorePaths": [
        "**/node_modules/**",
        "**/vscode-extension/**",
        "**/.git/**",
        "**/.pnpm-lock.json",
        ".vscode",
        "megalinter",
        "package-lock.json",
        "report"
    ],
    "language": "en",
    "noConfigSearch": true,
    "words": [
        "megalinter",
        "oxsecurity",
        "hadolint",
        "jsonlint",
        "isort",
        "mypy",
        "pylint",
        "pyright",
        "devskim",
        "dustilock",
        "gitleaks",
        "syft",
        "trivy"
    ],
    "version": "0.2"
}
```

**Comment by Rhernandez513** [comment]
@polatengin instead of maintaining our own, feedback to the team that maintains Github Action is more prudent.  If there isn't a clear and easy way to update the version (upstream) then I'll unblock, we can follow up in the Parking Lot.

**Review by Rhernandez513 — APPROVED** [review]

**Comment by polatengin** [comment]
> There is another Warning at https://github.com/benchpress-team/benchpress/actions/runs/3316951186/jobs/5479305221#step:4:254

> Warning: The set-output command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

> This one is non-blocking, but should be remedied as soon as convenient.

yeah, I'm aware of that. but we're not using `set-output` to promote a value as the output of the step. it's just a warning tells us that, if we're using `set-output`, it'll stop working. here is the official documentation to output a value from a step; https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter

**Pull request merged**
